libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ForEachElement.cs

```diff
@@ -27,6 +27,7 @@ public class ForEachElement : Dialog, IDialogDependencies
         private const string IterationValue = "value";
         private const string ActionScopeState = "this.actionScopeState";
         private const string ForeachIndex = "dialog.foreachIndex";
+        private const string CachedItemsProperty = "dialog._cachedItems";
```

Comment:
Shouldn't this be 'this' so that if there are 2 foreach in the same dialog then the property will exist at the foreach level? IIRC (DialogMemoryScope.cs) the 'dialog' scope operates at the scope of a DialogContainer which is the entire containing adaptive dialog. You can verify this by adding a tests with 2 foreach (nested perhaps) and checking if the arrays override each other.

Reply:
Changed to this.cachedItems, and also removed the ForeachIndex constant as it is a duplicate of the Index property field.